### PR TITLE
Merge 2.12.x to 2.13.x [ci: last-only]

### PR DESCRIPTION
**Pull request by retronym**
Only change include here is the optimizations to immutable.Map*.hashCode.

This PR ignores that change (`git merge -sours`). #8446 reworks this code for 2.13.x.

### CODE REVIEW

**Comment by retronym** [comment]
Marking as WIP until I do some more testing.

**Comment by retronym** [comment]
Addressed a few problems:

  - boxing of the hashcode coming out of the lambda argument to `nonAllocating { ... }` was costing an allocation. Not sure why the 2.12.x version of this test wasn't failing, maybe it was relying on JVM escape analysis which for some reason doesn't trigger in 2.13.x? Addressed here by using discarding the value. It occurs to me that we really ought to use `jmh.BlackHole` or similar to make sure that code under test is actually run, but I'll leave that for a separate change.
  - Map hash codes changed in 2.13.x because of https://github.com/scala/scala/pull/7634 and https://github.com/scala/scala/pull/7693. I've added a unit test to ensure consistency between the hand-inlined non-allocating versions and the generic version.

**Comment by mkeskells** [comment]
I modified the `AllocationTest`, e.g. `def nonAllocating` in the 2.12 version
it seems that due to the way that the lambda is captured the method under test generating a int, and that being boxed before checking, so that counted as an extra object

I changed the nonAllocation verson in 2.12 to take account of this allocation like this
```
  val costInt: Long = { 
    object coster extends AllocationTest 
    val costs = coster.allocationInfoImpl(coster.hashCode(), new AllocationExecution(), 0) 
    costs.min 
  } 
```

this wasn't in the 2.13 version, a we never tested allocation on a primitive before

**Comment by retronym** [comment]
I've pushed a commit to rework things further.

   - Forward port @mkeskells's changes to `AllocationTest` to account for generic boxing of `Integers`.
   - Centralize optimization in `MurmurHash3.mapHash` in terms of `Map.foreachEntry`.

File: src/testkit/scala/tools/testkit/AllocationTest.scala [hunk 1]
```diff
@@ -20,57 +20,81 @@ object AllocationTest {
...
+  println(s"cost of tracking allocations = $costObject")
 }
 
 trait AllocationTest {
 
   import AllocationTest._
 
   def nonAllocatingEqual(expected: Boolean, a: AnyRef, b: AnyRef): Unit = {
-    assertEquals(expected, nonAllocating(a == b))
+    assertEquals(expected, nonAllocating(if (a == b) "t" else "f"))
```
Comment (mkeskells) [inline src/testkit/scala/tools/testkit/AllocationTest.scala]: I think this is the cause of the failures in the last run
"t" cant equal true

File: src/testkit/scala/tools/testkit/AllocationTest.scala [hunk 2]
```diff
@@ -80,5 +104,13 @@ trait AllocationTest {
 case class AllocationExecution(executionCount: Int = 1000, warmupCount: Int = 1000)
 
 case class AllocationInfo[T](result: T, allocations: Array[Long]) {
-  def min = allocations.iterator.min
+  def min: Long = {
```
Comment (mkeskells) [inline src/testkit/scala/tools/testkit/AllocationTest.scala]: Note - I just unrolled this when looking for the extra allocations

File: src/testkit/scala/tools/testkit/AllocationTest.scala [hunk 3]
```diff
@@ -20,57 +20,81 @@ object AllocationTest {
...
+      i += 1
     }
 
-   //test
+    //test
+    i = 0
     val counts = new Array[Long](execution.executionCount)
-    for (i <- 0 until execution.executionCount) {
-      val before = allocationCounter.getThreadAllocatedBytes(id)
+    while (i < execution.executionCount) {
```
Comment (mkeskells) [inline src/testkit/scala/tools/testkit/AllocationTest.scala]: Note - I just unrolled this when looking for the extra allocations

File: build.sbt [hunk 4]
```diff
@@ -150,6 +150,10 @@ val mimaFilterSettings = Seq {
     ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.VectorPointer.gotoPosWritable1$default$4"),
     ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.VectorPointer.nullSlotAndCopy$default$3"),
     ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.mutable.HashMap.mapValuesInPlaceImpl"),
+    ProblemFilters.exclude[MissingClassProblem]("scala.collection.immutable.MapKeyValueTupleHashIterator"),
+    ProblemFilters.exclude[MissingClassProblem]("scala.collection.immutable.Map$HashCodeAccumulator"),
```
Comment (lrytz) [inline build.sbt]: This class seems to be unused in the final version

**Review by lrytz — CHANGES_REQUESTED** [review]

**Comment by lrytz** [comment]
Maybe it's more clear to merge with `-s ours` and apply the 2.13 changes separately?

**Comment by retronym** [comment]
I've made this PR into a `-sours` no-op merge and will submit a separate PR with the 2.13.x rework.

**Pull request merged**